### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) for releases

### DISCUSSION THREAD

**Discussion by zamazan4ik** [ChainSafe/forest#4858] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
...
lto = true
```

I see that for the quick profile you explicitly disable LTO - that's completely fine. I am just asking about enabling LTO for the Release-like configurations. If you think that LTO (even its Thin version) is too heavy for the compiler - then we can create a dedicated profile, smth like:

```
[profile.release-lto]
inherits = "release"
lto = true
```

Thank you.

**Comment by LesnyRumcajs** [comment]
Hey @zamazan4ik, thanks a lot for this. I was pretty surprised to learn that we don't have it enabled! I'm definitely open to enabling `thin` LTO, the `fat` one only in a separate profile given that it takes 4x more to build from scratch on my rather beefy machine (mostly linking) than the regular one, ~120s -> 430s.

**Reply by LesnyRumcajs** [reply]
This should be addressed in https://github.com/ChainSafe/forest/pull/4869.
